src/Rules/Regexp/RegularExpressionQuotingRule.php

```diff
@@ -0,0 +1,217 @@
+	 */
+	private function validateQuoteDelimiters(Concat $concat, Scope $scope, ?string $patternDelimiter): array
+	{
+		if ($patternDelimiter === null) {
+			return [];
+		}
+
+		$errors = [];
+		if (
+			$concat->left instanceof FuncCall
+			&& $concat->left->name instanceof Name
+			&& $concat->left->name->toLowerString() === 'preg_quote'
+		) {
+			$pregError = $this->validatePregQuote($concat->left, $scope, $patternDelimiter);
+			if ($pregError !== null) {
+				$errors[] = $pregError;
+			}
+		} elseif ($concat->left instanceof Concat) {
+			$errors = array_merge($errors, $this->validateQuoteDelimiters($concat->left, $scope, $patternDelimiter));
+		}
+
+		if (
+			$concat->right instanceof FuncCall
+			&& $concat->right->name instanceof Name
+			&& $concat->right->name->toLowerString() === 'preg_quote'
+		) {
+			$pregError = $this->validatePregQuote($concat->right, $scope, $patternDelimiter);
+			if ($pregError !== null) {
+				$errors[] = $pregError;
+			}
+		} elseif ($concat->right instanceof Concat) {
+			$errors = array_merge($errors, $this->validateQuoteDelimiters($concat->right, $scope, $patternDelimiter));
+		}
+
+		return $errors;
+	}
+
+	private function validatePregQuote(FuncCall $pregQuote, Scope $scope, string $patternDelimiter): ?IdentifierRuleError
+	{
+		if (!$pregQuote->name instanceof Node\Name) {
+			return null;
+		}
+
+		if (!$this->reflectionProvider->hasFunction($pregQuote->name, $scope)) {
+			return null;
+		}
+		$functionReflection = $this->reflectionProvider->getFunction($pregQuote->name, $scope);
+
+		$normalizedArgs = $this->getNormalizedArgs($pregQuote, $scope, $functionReflection);
+		if ($normalizedArgs === null) {
+			return null;
+		}
+
+		if (!isset($normalizedArgs[1])) {
+			return RuleErrorBuilder::message(sprintf('Call to preg_quote() is missing delimiter %s to be effective.', $patternDelimiter))
+				->line($pregQuote->getStartLine())
+				->identifier('argument.invalidPregQuote')
+				->build();
+		}
+
+		if (count($normalizedArgs) >= 2) {
+			foreach ($scope->getType($normalizedArgs[1]->value)->getConstantStrings() as $quoteDelimiterType) {
+				if ($quoteDelimiterType->getValue() !== $patternDelimiter) {
+					return RuleErrorBuilder::message(sprintf('Call to preg_quote() uses invalid delimiter %s while pattern uses %s.', $quoteDelimiterType->getValue(), $patternDelimiter))
+						->line($pregQuote->getStartLine())
+						->identifier('argument.invalidPregQuote')
+						->build();
+				}
+			}
+		}
+
+		return null;
+	}
+
+	/**
+	 * Get delimiters from non-constant patterns, if possible.
+	 */
+	private function getDelimiterFromConcat(Concat $concat, Scope $scope): ?string
+	{
+		if ($concat->left instanceof Concat) {
+			return $this->getDelimiterFromConcat($concat->left, $scope);
+		}
+
+		$left = $scope->getType($concat->left);
+
+		$constantStrings = $left->getConstantStrings();
+		// since we analyze a Expr directly given to a call, it can only ever resolve to a single constant string
```

Comment:
You bet it can be multiple strings. It should be handled here.

What about `$fooOrBar . preg_quote($s, '#') . $barOrBaz`.

Where `$fooOrBar` is `'foo'|'bar`.